// *****************************************************************************
//
// This file is part of a MASA library or program.
// Refer to the included end-user license agreement for restrictions.
//
// Copyright (c) 2010 MASA Group
//
// *****************************************************************************

package MsgsAuthenticationToClient;

import "Common.proto"; 
import "Version.proto";

message MsgAuthenticationResponse {
    enum ErrorCode {
        success                     = 0;
        invalid_login               = 1;
        too_many_connections        = 2;
        mismatched_protocol_version = 4;
    }
    required ErrorCode                 error_code        = 1;
    optional MsgProfile                profile           = 2; 
    optional MsgProfileDescriptionList profiles          = 3;
    required Version.ProtocolVersion   server_version  	 = 4;
	optional Common.MsgDateTime        restart_date_time = 5;
    optional string                    terrain_name      = 6;
}

message MsgProfileCreation {
    required MsgProfile profile = 1;
}

message MsgProfileCreationRequestAck {
    enum ErrorCode {
        success          = 0;
        invalid_login    = 1;
        invalid_password = 2;
        duplicate_login  = 3;
    }
    required string    login      = 1;
    required ErrorCode error_code = 2;
}

message MsgProfileUpdateRequestAck {
    enum ErrorCode {
        success          = 0;
        invalid_profile  = 1;
        invalid_login    = 2;
        invalid_password = 3;
        duplicate_login  = 4;
    }
    required string    login      = 1;
    required ErrorCode error_code = 2;
    optional int32 context        = 3;
}

message MsgProfileUpdate {
    required string     login   = 1;
    required MsgProfile profile = 2;
}

message MsgProfileDestruction {
    optional string login = 1;
}

message MsgProfileDestructionRequestAck {
    enum ErrorCode {
        success         = 0;
        invalid_profile = 1;
        failure         = 2;
    }
    required ErrorCode error_code = 1;
    required string    login      = 2;
}

message RoleType { required uint32 id = 1; }

message MsgProfile {
    required string                  login                   = 1;
    optional string                  password                = 2;
    optional Common.FormationIdList  read_only_formations    = 3;
    optional Common.FormationIdList  read_write_formations   = 4;
    optional Common.AutomatIdList    read_only_automates     = 5;
    optional Common.AutomatIdList    read_write_automates    = 6;
    optional Common.PartyIdList      read_only_camps         = 7;
    optional Common.PartyIdList      read_write_camps        = 8;
    optional Common.CrowdIdList      read_only_crowds        = 9;
    optional Common.CrowdIdList      read_write_crowds       = 10;
    optional Common.PopulationIdList write_population        = 11;
    required bool                    supervisor              = 12;
    optional RoleType                role                    = 13;
}

message MsgProfileDescriptionList {
    repeated MsgProfileDescription elem = 1;
}

message MsgProfileDescription {
    required string   login      = 1;
    required bool     password   = 2;
    required bool     supervisor = 3;
    optional RoleType role       = 4;
}

message MsgAuthenticationToClient {
    message Content {
        optional MsgAuthenticationResponse       authentication_response         = 1;
        optional MsgProfileCreation              profile_creation                = 2;
        optional MsgProfileCreationRequestAck    profile_creation_request_ack    = 3;
        optional MsgProfileUpdate                profile_update                  = 4;
        optional MsgProfileUpdateRequestAck      profile_update_request_ack      = 5;
        optional MsgProfileDestruction           profile_destruction             = 6;
        optional MsgProfileDestructionRequestAck profile_destruction_request_ack = 7;
    }
    optional int32   context = 1;
    required Content message = 2;
}
